File: langs/swift/Sources/VMx/Composites/FilteredCompositeVM.swift
//
// FilteredCompositeVM.swift — visible projection over a CompositeVM source.
//
import Combine

public enum FilteredCursorPolicy {
    case snapToFirst
    case clear
    case preserveIfVisible
}

open class FilteredCompositeVM<Child: ComponentVMBase> {
    let source: CompositeVM<Child>
    private var predicate: (Child) -> Bool
    private let cursorPolicy: FilteredCursorPolicy
    private var visibleStorage: [Child] = []
    private var cancellables: Set<AnyCancellable> = []
    private var disposed = false
    private let changedSubject = PassthroughSubject<Void, Never>()

    public var visible: [Child] { visibleStorage }
    public var visibleCount: Int { visibleStorage.count }
    public var current: Child?

    public var changed: AnyPublisher<Void, Never> {
        changedSubject.eraseToAnyPublisher()
    }

    public init(
        _ source: CompositeVM<Child>,
        cursorPolicy: FilteredCursorPolicy = .snapToFirst,
        predicate: @escaping (Child) -> Bool = { _ in true }
    ) {
        self.source = source
        self.cursorPolicy = cursorPolicy
        self.predicate = predicate
        source.collectionChanged
            .sink { [weak self] _ in self?.recompute() }
            .store(in: &cancellables)
        recompute()
    }

    init(
        _ source: CompositeVM<Child>,
        cursorPolicy: FilteredCursorPolicy,
        predicate: @escaping (Child) -> Bool,
        deferInitialRecompute: Bool
    ) {
        self.source = source
        self.cursorPolicy = cursorPolicy
        self.predicate = predicate
        source.collectionChanged
            .sink { [weak self] _ in self?.recompute() }
            .store(in: &cancellables)
        if !deferInitialRecompute { recompute() }
    }

    public func setPredicate(_ predicate: @escaping (Child) -> Bool) {
        self.predicate = predicate
        recompute()
    }

    public func setCurrent(_ item: Child?) {
        if let item, !visibleStorage.contains(where: { $0 === item }) {
            preconditionFailure("current must be nil or a visible item")
        }
        if current === item { return }
        current = item
        changedSubject.send(())
    }

    public func moveToNextVisible() {
        guard !visibleStorage.isEmpty else {
            setCurrent(nil)
            return
        }
        guard let current, let index = visibleStorage.firstIndex(where: { $0 === current }) else {
            setCurrent(visibleStorage[0])
            return
        }
        setCurrent(visibleStorage[min(index + 1, visibleStorage.count - 1)])
    }

    public func moveToPreviousVisible() {
        guard !visibleStorage.isEmpty else {
            setCurrent(nil)
            return
        }
        guard let current, let index = visibleStorage.firstIndex(where: { $0 === current }) else {
            setCurrent(visibleStorage[0])
            return
        }
        setCurrent(visibleStorage[max(index - 1, 0)])
    }

    open func orderedVisible() -> [Child] {
        (0..<source.count).map { source.at($0) }.filter(predicate)
    }

    func recompute() {
        visibleStorage = orderedVisible()
        if let current, !visibleStorage.contains(where: { $0 === current }) {
            self.current = cursorPolicy == .snapToFirst ? visibleStorage.first : nil
        } else if current == nil && cursorPolicy == .snapToFirst {
            current = visibleStorage.first
        }
        changedSubject.send(())
    }

    public func dispose() {
        guard !disposed else { return }
        disposed = true
        cancellables.removeAll()
        changedSubject.send(completion: .finished)
    }
}

File: langs/typescript/src/composites/filteredCompositeVM.ts
/**
 * FilteredCompositeVM — visible projection over a CompositeVM source.
 */
import { Subject, Subscription } from "rxjs";
import type { Observable } from "rxjs";
import { ComponentVMBase } from "../components/componentVMBase.js";
import type { CompositeVMBase } from "./compositeVMBase.js";

export enum FilteredCursorPolicy {
  SnapToFirst = "snapToFirst",
  Clear = "clear",
  PreserveIfVisible = "preserveIfVisible",
}

export interface FilteredCompositeOptions<VM extends ComponentVMBase> {
  readonly predicate?: (vm: VM) => boolean;
  readonly cursorPolicy?: FilteredCursorPolicy;
  readonly deferInitialRecompute?: boolean;
}

export class FilteredCompositeVM<VM extends ComponentVMBase> {
  protected readonly source: CompositeVMBase<VM>;
  protected predicate: (vm: VM) => boolean;
  readonly #cursorPolicy: FilteredCursorPolicy;
  readonly #changed = new Subject<void>();
  readonly #subscription: Subscription;
  #visible: VM[] = [];
  #current: VM | null = null;
  #disposed = false;

  constructor(
    source: CompositeVMBase<VM>,
    options: FilteredCompositeOptions<VM> = {},
  ) {
    this.source = source;
    this.predicate = options.predicate ?? (() => true);
    this.#cursorPolicy = options.cursorPolicy ?? FilteredCursorPolicy.SnapToFirst;
    this.#subscription = source.collectionChanged.subscribe(() => this._recompute());
    if (options.deferInitialRecompute !== true) this._recompute();
  }

  get visible(): VM[] {
    return [...this.#visible];
  }

  get visibleCount(): number {
    return this.#visible.length;
  }

  get current(): VM | null {
    return this.#current;
  }

  set current(value: VM | null) {
    this.setCurrent(value);
  }

  get changed(): Observable<void> {
    return this.#changed.asObservable();
  }

  setPredicate(predicate: (vm: VM) => boolean): void {
    this.predicate = predicate;
    this._recompute();
  }

  setCurrent(item: VM | null): void {
    if (item !== null && !this.#visible.includes(item)) {
      throw new Error("current must be null or a visible item");
    }
    if (this.#current === item) return;
    this.#current = item;
    this.#changed.next();
  }

  moveToNextVisible(): void {
    if (this.#visible.length === 0) {
      this.setCurrent(null);
      return;
    }
    if (this.#current === null || !this.#visible.includes(this.#current)) {
      this.setCurrent(this.#visible[0] ?? null);
      return;
    }
    const index = this.#visible.indexOf(this.#current);
    this.setCurrent(this.#visible[Math.min(index + 1, this.#visible.length - 1)] ?? null);
  }

  moveToPreviousVisible(): void {
    if (this.#visible.length === 0) {
      this.setCurrent(null);
      return;
    }
    if (this.#current === null || !this.#visible.includes(this.#current)) {
      this.setCurrent(this.#visible[0] ?? null);
      return;
    }
    const index = this.#visible.indexOf(this.#current);
    this.setCurrent(this.#visible[Math.max(index - 1, 0)] ?? null);
  }

  protected orderedVisible(): VM[] {
    return [...this.source].filter((vm) => this.predicate(vm));
  }

  protected _recompute(): void {
    this.#visible = this.orderedVisible();
    if (this.#current !== null && !this.#visible.includes(this.#current)) {
      this.#current =
        this.#cursorPolicy === FilteredCursorPolicy.SnapToFirst
          ? (this.#visible[0] ?? null)
          : null;
    } else if (this.#current === null && this.#cursorPolicy === FilteredCursorPolicy.SnapToFirst) {
      this.#current = this.#visible[0] ?? null;
    }
    this.#changed.next();
  }

  dispose(): void {
    if (this.#disposed) return;
    this.#disposed = true;
    this.#subscription.unsubscribe();
    this.#changed.complete();
  }
}
